feat: `avm/res/document-db/database-account` Add the creation of 'Microsoft.DocumentDB/databaseAccounts/sqlRoleDefinitions' and ‘Microsoft.DocumentDB/databaseAccounts/sqlRoleAssignments’ and 'cosmos.properties.documentEndpoint'.

zedy-wj opened the pull request:
## Description

<!--
>Thank you for your contribution !
> Please include a summary of the change and which issue is fixed.
> Please also include the context.
> List any dependencies that are required for this change.

Fixes #123
Fixes #456
Closes #123
Closes #456
-->

Fixes #1126 
## Pipeline Reference

<!-- Insert your Pipeline Status Badge below -->

| Pipeline |
| -------- |
|[![avm.res.document-db.database-account](https://github.com/zedy-wj/bicep-registry-modules/actions/workflows/avm.res.document-db.database-account.yml/badge.svg?branch=fix%2F1126)](https://github.com/zedy-wj/bicep-registry-modules/actions/runs/9481159887/job/26165484398)          |

The failure in the pipeline is due to the region hardcoded as `eastasia` in the test files, and in our subscription, there is no permission to create storage in this region. The region in the newly added test file for `sql-role` has been changed to eastus2 and it passed.
## Type of Change

<!-- Use the check-boxes [x] on the options that are relevant. -->

- [ ] Update to CI Environment or utlities (Non-module effecting changes)
- [ x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [ x] I'm sure there are no other open Pull Requests for the same update/change
- [ x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [ x] My corresponding pipelines / checks run clean and green without any errors or warnings

<!--  Please keep up to day with the contribution guide at https://aka.ms/avm/contribute/bicep -->

@jongio for notification.

bryansan-msft commented on avm/res/document-db/database-account/tests/e2e/role/dependencies.bicep:
```diff
@@ -0,0 +1,77 @@
+@description('Optional. Managed service identity.')
+param managedIdentity bool = !empty(keyVaultName)
+
+resource keyVault 'Microsoft.KeyVault/vaults@2021-06-01-preview' = {
```
This key vault is not used for testing. Can we remove it?

bryansan-msft commented on avm/res/document-db/database-account/tests/e2e/role/dependencies.bicep:
```diff
@@ -0,0 +1,77 @@
+param appServicePlanReserved bool = true
+
+@description('Optional. Managed service identity.')
+param managedIdentity bool = !empty(keyVaultName)
```
Your test always needs a managed identity. Can we remove this parameter and always emit an identityPrincipalId without a condition?

bryansan-msft commented on avm/res/document-db/database-account/tests/e2e/role/main.test.bicep:
```diff
@@ -0,0 +1,60 @@
+targetScope = 'subscription'
+
+metadata name = 'Deploying with a key vault reference to save secrets'
```
This metadata value is a copy paste of another test. Can we write a meaningful one?

bryansan-msft commented on avm/res/document-db/database-account/tests/e2e/role/dependencies.bicep:
```diff
@@ -0,0 +1,77 @@
+  'app,linux' // linux web app
+  'app' // normal web app
+])
+param appKind string = 'app,linux'
```
Can we simplify all of these parameters and just hardcode them in the bicep resource directly? your test is not using them and this is just added extra noise

bryansan-msft commented on avm/res/document-db/database-account/sql-role/sql-role-definitions/main.bicep:
```diff
@@ -0,0 +1,53 @@
+
+@description('Optional. An array of data actions that are allowed.')
+param dataActions array = [
+  'Microsoft.DocumentDB/databaseAccounts/readMetadata'
```
Can we decide whether these default values are here or in the parent "sql-role/main.bicep" file? These default values are duplicated these two files.

bryansan-msft commented on avm/res/document-db/database-account/sql-role/main.bicep:
```diff
@@ -0,0 +1,66 @@
+
+@description('Optional. An array of data actions that are allowed.')
+param dataActions array = [
+  'Microsoft.DocumentDB/databaseAccounts/readMetadata'
```
As i mentioned in my other comment. We need to either have defaults value here or in its children files but not in both as those defaults values are duplicated and generated extra noise and confusion

bryansan-msft reviewed (changes requested):
Added comments mostly in code simplification regarding unused parameters

zedy-wj commented:
> Added comments mostly in code simplification regarding unused parameters. Can you please link the workflow run with your new role test in green? the link you posted in the PR description is not opening for me

We have updated  the link to the new role test workflow in the description, and also updated the code based on your suggestion, please review. Thanks.

bryansan-msft reviewed (approved):
Looks good to me
